### PR TITLE
Add NEWS entry for adjtrans of Triangular types

### LINKED ISSUE #783: Method resolution issue for `rdiv!`
The change in JuliaLang/julia#38168 causes MixedModels to fail on Julia master with a method resolution error for `rdiv!`:

```julia
  MethodError: no method matching rdiv!(::BlockedSparse{Float64, 1, 2}, ::UpperTriangular{Float64, Adjoint{Float64, UniformBlo
ckDiagonal{Float64}}})                                         
  Closest candidates are:
    rdiv!(::StridedMatrix{T} where T, ::UpperTriangular{var"#s814", var"#s813"} where var"#s813"<:Adjoint where var"#s814") at  ~/julia/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/triangular.jl:1430
    rdiv!(::StridedMatrix{T} where T, ::UpperTriangular) at ~/julia/usr/share/julia/stdlib/v1.6/LinearAlgebra
/src/triangular.jl:1327
    rdiv!(::AbstractMatrix{T} where T, ::Transpose{var"#s826", var"#s825"} where var"#s825"<:LU where var"#s826") at ~/julia/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/lu.jl:689
```

Tests pass again after reverting this change in a local build.

`BlockedSparse` is a wrapper around `SparseMatrixCSC` with a particular sparsity pattern. Note that this does **not** fall back to the generic multiplication method -- the change in order means that the second argument doesn't match the second argument in the last of the closest candidates.

### LINKED ISSUE #38378: Method resolution issue for `rdiv!`
The change in #38168 causes MixedModels to fail on Julia master with a method resolution error for `rdiv!`:

```julia
  MethodError: no method matching rdiv!(::BlockedSparse{Float64, 1, 2}, ::UpperTriangular{Float64, Adjoint{Float64, UniformBlo
ckDiagonal{Float64}}})                                         
  Closest candidates are:
    rdiv!(::StridedMatrix{T} where T, ::UpperTriangular{var"#s814", var"#s813"} where var"#s813"<:Adjoint where var"#s814") at  ~/julia/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/triangular.jl:1430
    rdiv!(::StridedMatrix{T} where T, ::UpperTriangular) at ~/julia/usr/share/julia/stdlib/v1.6/LinearAlgebra
/src/triangular.jl:1327
    rdiv!(::AbstractMatrix{T} where T, ::Transpose{var"#s826", var"#s825"} where var"#s825"<:LU where var"#s826") at ~/julia/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/lu.jl:689
```

Tests pass again after reverting this change in a local build.

`BlockedSparse` is a wrapper around `SparseMatrixCSC` with a particular sparsity pattern. Note that this does **not** fall back to the generic multiplication method -- the change in order means that the second argument doesn't match the second argument in the last of the closest candidates.

### PR DESCRIPTION
**Pull request by dkarrasch**
This is a follow up to JuliaLang/julia#38168, adding a NEWS entry, removing dead/commented out code, and adding type tests.

Ping @palday. 😉 

Closes JuliaLang/LinearAlgebra.jl#783.

### CODE REVIEW

**Comment by palday** [comment]
I'll review tomorrow -- I need to think a little bit about whether the tests are covering all the cases I can think of and don't have the mental capacity today.

The NEWS looks good.

**Comment by dkarrasch** [comment]
The tests don't check the type explicitly, but only consistency of the type name (not the parameters) between `adjoint!` and `adjoint`. I was too lazy to write some loop with complementary pairs of types etc. Arguably, this consistency can be considered as the "bug fix" part of the original PR.

**Comment by dkarrasch** [comment]
While this doesn't have new code, can we still launch a package evaluation, to see who to inform about the change in #38168? @StefanKarpinski @ViralBShah

**Comment by ViralBShah** [comment]
Yes, we can. Perhaps @maleadt can tell us how to trigger PkgEval on before and after #38168.

**Comment by palday** [comment]
> The tests don't check the type explicitly, but only consistency of the type name (not the parameters) between adjoint! and adjoint. I was too lazy to write some loop with complementary pairs of types etc. 

I get that. :smile: 

> Arguably, this consistency can be considered as the "bug fix" part of the original PR.

Yep, seeing it like this I can understand the motivation of the original change better. The actual order of the types seems arbitrary-ish to me (or rather I can see arguments for both orders), but consistency is important.

File: NEWS.md [hunk 1]
```diff
@@ -164,6 +164,10 @@ Standard library changes
 * `(+)(::UniformScaling)` is now defined, making `+I` a valid unary operation ([#36784]).
 * Instances of `UniformScaling` are no longer `isequal` to matrices. Previous
   behaviour violated the rule that `isequal(x, y)` implies `hash(x) == hash(y)`.
+* Transposing `*Triangular` matrices now returns matrices of the opposite triangular type, consistently
```
Comment (palday) [inline NEWS.md:167]:
```suggestion
* Transposing `*Triangular` matrices now returns matrices of the opposite triangular type, consistent
```

It is an adverbial use, but `consistent` sounds better than `consistently` for me here.

(I'm technically a native speaker, but I would not depend on my intuitions.)

**Review by palday — APPROVED** [review]

**Comment by maleadt** [comment]
PkgEval running here: https://github.com/JuliaLang/julia/commit/4704efd6bdb6fe2ace913055ef66bd52f8da8fac#commitcomment-44115567

File: NEWS.md (hunk 1 above)
Comment (ViralBShah) [reply NEWS.md:167]: I believe the original is right, and also sounds better.

**Comment by dkarrasch** [comment]
Thanks @maleadt! AFAICT, the `MixedModels.jl` failure is the only one induced by the original PR.

**Comment by ViralBShah** [comment]
Results of the PkgEval run- https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/4704efd_vs_19506f7/report.md

**Comment by ararslan** [comment]
> AFAICT, the MixedModels.jl failure is the only one induced by the original PR.

The ChainRules failure is likely related, as it relies heavily on specialization with LinearAlgebra types.

**Comment by dkarrasch** [comment]
> > AFAICT, the MixedModels.jl failure is the only one induced by the original PR.
> 
> The ChainRules failure is likely related, as it relies heavily on specialization with LinearAlgebra types.

I couldn't even find the term `Triangular` in their code base. But let's see. @oxinabox

**Comment by oxinabox** [comment]
ChainRules uses Triangular matrixes in
https://github.com/JuliaDiff/ChainRules.jl/blob/d40e3e174524a6f40e40b114e6e89d3bd413c6d3/src/rulesets/LinearAlgebra/structured.jl
and it calls `adjoint` on them.
Quite possibly it should just be calling `conj` on them.
I would need to think a bit I guess.

**Comment by dkarrasch** [comment]
There is only one error in ChainRules.jl:
```julia
3-arg dot, Array{ComplexF64}: Test Failed at /home/pkgeval/.julia/packages/ChainRulesTestUtils/wjZiE/src/check_result.jl:18
  Expression: isapprox(add!!(acc_mutated, val), acc + val; kwargs...)
   Evaluated: isapprox(ComplexF64[-1.1701131866419396 - 3.79687561601824im, -0.36934589076936375 - 1.7677498974545742im, 2.9343957655656716 + NaN*im], ComplexF64[-1.1701131866419396 - 3.79687561601824im, -0.36934589076936375 - 1.7677498974545742im, 2.9343957655656716 + NaN*im]; rtol = 1.0e-9, atol = 1.0e-9)
```
It looks to me like an accuracy issue, not a deep type-related failure. What is potentially critical for downstream packages is when they methods with signatures à la `Adjoint{<:Any,<:LowerTriangular}`, because these types are no longer realized (instead `UpperTriangular{<:Any,<:Adjoint}`). This is what happened in `MixedModels.jl`.

**Comment by oxinabox** [comment]
Oh that. That is not related to this.
It's a no deterministic fault because some tests just apparently assumed for `y=similar(x)` that `isapprox(y, y)`
It was fixed the other day
https://github.com/JuliaDiff/ChainRules.jl/pull/304

**Comment by dkarrasch** [comment]
Are we happy with this? Ready to go?

**Pull request merged**